### ISSUE TITLE
[M1 macOS] Help wanted for testing Lima with QEMU@master

### ISSUE THREAD

**Issue by AkihiroSuda** [lima-vm/lima#713]
Help wanted for testing https://github.com/lima-vm/lima/pull/703 on M1 macOS

1. Install Lima v0.9.0-beta.0 or later (`brew install --HEAD lima`, or `git clone https://github.com/lima-vm/lima.git && cd lima && make && make install`)
2. Install QEMU v6.2.0 (`brew install qemu`)
3. Create instances with the following memory: 8 GiB, 4 GiB, 2 GiB.
```
$ limactl start
? Creating an instance "default"  [Use arrows to move, type to filter]
  Proceed with the current configuration
> Open an editor to review or modify the current configuration
  Choose another example (docker, podman, archlinux, fedora, ...)
  Exit

(Change "memory: null" to "memory: 8 GiB" in the editor, and confirm that the instance works)
$ limactl delete -f defaut
```
4. Install QEMU master (`brew install --HEAD qemu`)
5. Repeat the step 3

**Comment by Junnplus** [comment]
I work on M1 macOS, I will take a look.

**Comment by Junnplus** [comment]
```
[N] ❯ limactl --version
limactl version HEAD-3a6502a

[N] ❯ qemu-system-aarch64 --version
QEMU emulator version 6.2.0
Copyright (c) 2003-2021 Fabrice Bellard and the QEMU Project developers
```

Work successfully on 8 GiB, 4 GiB, 2 GiB.
```
[N] ❯ ps aux|grep qemu
jun              90698   1.7  9.6 412361008 3232432 s021  S    10:23PM   1:13.15 /opt/homebrew/bin/qemu-system-aarch64 -cpu host -machine virt,accel=hvf,highmem=off -smp 4,sockets=1,cores=4,threads=1 -m 2048 -drive if=pflash,format=raw,readonly=on,file=/opt/homebrew/share/qemu/edk2-aarch64-code.fd -boot order=c,splash-time=0,menu=on -drive file=/Users/jun/.lima/test2/diffdisk,if=virtio -cdrom /Users/jun/.lima/test2/cidata.iso -netdev user,id=net0,net=192.168.5.0/24,dhcpstart=192.168.5.15,hostfwd=tcp:127.0.0.1:62708-:22 -device virtio-net-pci,netdev=net0,mac=52:55:55:7f:52:9d -device virtio-rng-pci -display none -vga none -device ramfb -device qemu-xhci,id=usb-bus -device usb-kbd,bus=usb-bus.0 -device usb-mouse,bus=usb-bus.0 -parallel none -chardev socket,id=char-serial,path=/Users/jun/.lima/test2/serial.sock,server=on,wait=off,logfile=/Users/jun/.lima/test2/serial.log -serial chardev:char-serial -chardev socket,id=char-qmp,path=/Users/jun/.lima/test2/qmp.sock,server=on,wait=off -qmp chardev:char-qmp -name lima-test2 -pidfile /Users/jun/.lima/test2/qemu.pid
jun              90439   1.4  9.5 414432144 3184464 s020  S    10:22PM   1:13.54 /opt/homebrew/bin/qemu-system-aarch64 -cpu host -machine virt,accel=hvf,highmem=off -smp 4,sockets=1,cores=4,threads=1 -m 4096 -drive if=pflash,format=raw,readonly=on,file=/opt/homebrew/share/qemu/edk2-aarch64-code.fd -boot order=c,splash-time=0,menu=on -drive file=/Users/jun/.lima/test1/diffdisk,if=virtio -cdrom /Users/jun/.lima/test1/cidata.iso -netdev user,id=net0,net=192.168.5.0/24,dhcpstart=192.168.5.15,hostfwd=tcp:127.0.0.1:62662-:22 -device virtio-net-pci,netdev=net0,mac=52:55:55:9a:04:29 -device virtio-rng-pci -display none -vga none -device ramfb -device qemu-xhci,id=usb-bus -device usb-kbd,bus=usb-bus.0 -device usb-mouse,bus=usb-bus.0 -parallel none -chardev socket,id=char-serial,path=/Users/jun/.lima/test1/serial.sock,server=on,wait=off,logfile=/Users/jun/.lima/test1/serial.log -serial chardev:char-serial -chardev socket,id=char-qmp,path=/Users/jun/.lima/test1/qmp.sock,server=on,wait=off -qmp chardev:char-qmp -name lima-test1 -pidfile /Users/jun/.lima/test1/qemu.pid
jun              90047   0.3  9.9 418637776 3320944 s020  S    10:20PM   1:07.94 /opt/homebrew/bin/qemu-system-aarch64 -cpu host -machine virt,accel=hvf,highmem=off -smp 4,sockets=1,cores=4,threads=1 -m 8192 -drive if=pflash,format=raw,readonly=on,file=/opt/homebrew/share/qemu/edk2-aarch64-code.fd -boot order=c,splash-time=0,menu=on -drive file=/Users/jun/.lima/test/diffdisk,if=virtio -cdrom /Users/jun/.lima/test/cidata.iso -netdev user,id=net0,net=192.168.5.0/24,dhcpstart=192.168.5.15,hostfwd=tcp:127.0.0.1:62556-:22 -device virtio-net-pci,netdev=net0,mac=52:55:55:fa:1b:77 -device virtio-rng-pci -display none -vga none -device ramfb -device qemu-xhci,id=usb-bus -device usb-kbd,bus=usb-bus.0 -device usb-mouse,bus=usb-bus.0 -parallel none -chardev socket,id=char-serial,path=/Users/jun/.lima/test/serial.sock,server=on,wait=off,logfile=/Users/jun/.lima/test/serial.log -serial chardev:char-serial -chardev socket,id=char-qmp,path=/Users/jun/.lima/test/qmp.sock,server=on,wait=off -qmp chardev:char-qmp -name lima-test -pidfile /Users/jun/.lima/test/qemu.pid
```

**Comment by Junnplus** [comment]
```
[N] ❯ qemu-system-aarch64 --version
QEMU emulator version 6.2.50 (v6.2.0-2357-g2048c4eba2)
Copyright (c) 2003-2022 Fabrice Bellard and the QEMU Project developers
```

Unable to start the instance, the system crashes directly. :(

**Comment by AkihiroSuda** [comment]
Thank you @Junnplus for analysis.

> panic(cpu 3 caller 0xfffffe001c0f6a58): vm_fault() KERN_FAILURE from guest fault on state 0xfffffe6019168000 @sleh.c:3091

Could you report this to the QEMU upstream https://gitlab.com/qemu-project/qemu/-/issues , and also Apple, maybe?

**Comment by chancez** [comment]
I also got a panic.

```
(⎈ |rancher-desktop:default) ~ ❯❯❯ limactl --version                                                                                                                                                                                                                                                                                            ✘ 1 
limactl version HEAD-1164273

(⎈ |rancher-desktop:default) ~ ❯❯❯ qemu-system-aarch64 --version
QEMU emulator version 6.2.50 (v6.2.0-2380-g1416688c53)
Copyright (c) 2003-2022 Fabrice Bellard and the QEMU Project developers
```


```
panic(cpu 3 caller 0xfffffe001db6ea58): vm_fault() KERN_FAILURE from guest fault on state 0xfffffe6032c98000 @sleh.c:3091
Debugger message: panic
Memory ID: 0x6
OS release type: User
OS version: 21A559
Kernel version: Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:01 PDT 2021; root:xnu-8019.41.5~1/RELEASE_ARM64_T6000
```

I'll file a bug in qemu upstream and send my problem report to apple.

**Comment by chancez** [comment]
Qemu issue: https://gitlab.com/qemu-project/qemu/-/issues/903

**Comment by christhomas** [comment]
I've got an m1 max and I can't even run nerdctl cause it says that it's not installed and I'm unsure what to do next, I've filed a bug ticket about it

**Comment by Junnplus** [comment]
> I've got an m1 max and I can't even run nerdctl cause it says that it's not installed and I'm unsure what to do next, I've filed a bug ticket about it

It has nothing to do with this issue.

**Comment by AkihiroSuda** [comment]
> panic(cpu 3 caller 0xfffffe001db6ea58): vm_fault() KERN_FAILURE from guest fault on state 0xfffffe6032c98000 @sleh.c:3091

Is this still an issue with macOS 12.3?
Has anyone reported to Apple?

**Comment by jandubois** [comment]
I only have access to an M1 mini and not a pro/max chip, so can't test it myself (the qemu bug mentions that the M1 chip is unaffected)

**Comment by chancez** [comment]
Panic on 12.3 w/ qemu HEAD:

```
panic(cpu 2 caller 0xfffffe001546d814): vm_fault() KERN_FAILURE from guest fault on state 0xfffffe6032ae4000 @sleh.c:3117
Debugger message: panic
Memory ID: 0x6
OS release type: User
OS version: 21E230
Kernel version: Darwin Kernel Version 21.4.0: Mon Feb 21 20:35:58 PST 2022; root:xnu-8020.101.4~2/RELEASE_ARM64_T6000
Fileset Kernelcache UUID: 65F083231F3684E89C62E16ACD8BF64E
Kernel UUID: 178AA913-FCA9-33FD-A81C-DF08315F458D
```

**Comment by network-shark** [comment]
For me it works 👍 , but I did not update qemu6.2.0 to 6.2.0_1 .

╰─ limactl list
NAME       STATUS     SSH                ARCH       CPUS    MEMORY    DISK      DIR
default    Running    127.0.0.1:60022    aarch64    4       4GiB      100GiB    /Users/sharky/.lima/default

**Comment by network-shark** [comment]
How can I accomplish it ? 

> Thanks all for testing.
> Can anybody check whether https://lists.gnu.org/archive/html/qemu-devel/2022-03/msg00494.html fixes the host panic on M1 (Pro?)?

How can I use the patches ?

**Comment by chancez** [comment]
I'll test the patches. 

Here's how I'm testing:
```
brew unlink qemu
brew edit qemu
```

then remove all the patches , and modify the following line:
```
  head "https://git.qemu.org/git/qemu.git", branch: "master"
```
And update it to be:
```
  head "https://github.com/patchew-project/qemu.git", tag: "patchew/20220302130417.18551-1-ubzeme@gmail.com"
```

Then install the qemu like this:
```
export HOMEBREW_NO_AUTO_UPDATE=1
brew install --build-from-source --HEAD qemu
```

You can also just checkout the homebrew branch I created with these changes if you like:
https://github.com/chancez/homebrew-core/tree/qemu_hvf_patched


```
cd "$(brew --repository homebrew/core)"
git remote add chancez git@github.com:chancez/homebrew-core.git
git fetch chancez
git checkout chancez/qemu_hvf_patched
export HOMEBREW_NO_AUTO_UPDATE=1
brew unlink qemu
brew install --build-from-source --HEAD qemu
```

**Comment by chancez** [comment]
Got another panic with the above patches

```
(⎈ |rancher-desktop:default) ~ ❯❯❯ qemu-system-aarch64 --version
QEMU emulator version 6.2.90 (v6.2.0-2525-g1535ec6d77)
Copyright (c) 2003-2022 Fabrice Bellard and the QEMU Project developers
```

`v6.2.0-2525-g1535ec6d77` the last bit `g1535ec6d77`, without the leading g, matches the SHA here: https://github.com/patchew-project/qemu/commit/1535ec6d771df68200a85ce388e05ff0a34ffaa3

Here's the panic:

```
panic(cpu 3 caller 0xfffffe00146f5814): vm_fault() KERN_FAILURE from guest fault on state 0xfffffe61154e4000 @sleh.c:3117
Debugger message: panic
Memory ID: 0x6
OS release type: User
OS version: 21E230
Kernel version: Darwin Kernel Version 21.4.0: Mon Feb 21 20:35:58 PST 2022; root:xnu-8020.101.4~2/RELEASE_ARM64_T6000
Fileset Kernelcache UUID: 65F083231F3684E89C62E16ACD8BF64E
Kernel UUID: 178AA913-FCA9-33FD-A81C-DF08315F458D
iBoot version: iBoot-7459.101.2
```

**Comment by chancez** [comment]
According to another user, it sounds like this may be fixed in macOS 12.4 Beta1

https://gitlab.com/qemu-project/qemu/-/issues/903#note_911000975
